### ISSUE TITLE
Incorrect "variable is not accessed" message

### ISSUE THREAD

**Issue by jdayton3** [microsoft/pyright#8807]
<h3> Environment data </h3>

-   Language Server version: 2024.8.1
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): Python 3.8.10, in venv
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: strict

<h3> Code Snippet </h3>

```python
# pyright: strict
from functools import partial
from dataclasses import dataclass

@dataclass
class Dog:
    name: str
    breed: str

def partial_dog(breed: str):
    formatted_breed = breed.title() # error: Variable "formatted_breed" is not accessed

    return partial(
        Dog,
        breed=f"Maybe a {formatted_breed}", # but it's accessed right here
    )

if __name__ == "__main__":
    print(partial_dog("beagle")(name="Buddy"))
```

```powershell
C:\Users\jdayton\source\temp> python .\repro.py   
Dog(name='Buddy', breed='Maybe a Beagle')
```

<h3> Expected behavior </h3>

When using a function-local variable in an f-string as a `dataclass` constructor argument in `functools.partial`, Pylance should recognize that the variable is being accessed. 

<h3> Actual behavior </h3>

Pylance reports the function-local variable as unused.

We get the same error when running pyright (`pyright==1.1.377`, installed with pip) in the terminal:

```powershell
C:\Users\jdayton\source\temp> pyright .\repro.py
c:\Users\jdayton\source\temp\repro.py
  c:\Users\jdayton\source\temp\repro.py:11:5 - error: Variable "formatted_breed" is not accessed (reportUnusedVariable)
1 error, 0 warnings, 0 informations
```

We also get this error if `Dog` is a pydantic `BaseModel`:

```python
from pydantic import BaseModel # pydantic==2.7.4

class Dog(BaseModel):
    name: str
    breed: str
```

We **don't** get this error if `Dog` is a regular class:

```python
class Dog:
    def __init__(self, name: str, breed: str):
        self.name = name
        self.breed = breed
```

We also **don't** get the error if `formatted_breed` is passed in directly, not in an f-string:

```python
def partial_dog(breed: str):
    formatted_breed = f"Maybe a {breed.title()}"

    return partial(
        Dog,
        breed=formatted_breed,
    )
```

If the variable in the f-string is a parameter instead of a local variable, we don't get this error, but Pylance *sometimes* can't tell what its type is:

![image](https://github.com/user-attachments/assets/43db496c-8457-4baf-9f6e-547d3bb60eae)

**Comment by rchiodo** [comment]
Transferring to Pyright, typechecking issue.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.378.
